src/core/server/models/story/story.ts

```diff
@@ -911,3 +939,531 @@ export async function markStoryAsUnarchived(
+  comment: StoryTreeComment,
+  rootIndex: number,
+  result: FlattenedTreeComment[]
+) {
+  result.push({ ...comment, rootIndex });
+
+  // Default ordering in story tree is newest first, but replies are
+  // always oldest first order, so we need to flip this around.
+  for (let i = comment.replies.length - 1; i >= 0; i--) {
+    const reply = comment.replies[i];
+    flattenComment(reply, rootIndex, result);
+  }
+}
+
+function flattenTree(
+  tree: StoryTreeComment[],
+  orderBy: GQLCOMMENT_SORT,
+  result: FlattenedTreeComment[]
+) {
+  if (orderBy === GQLCOMMENT_SORT.CREATED_AT_DESC) {
+    for (let i = 0; i < tree.length; i++) {
+      const rootComment = tree[i];
+      flattenComment(rootComment, i, result);
+    }
+  } else if (orderBy === GQLCOMMENT_SORT.CREATED_AT_ASC) {
+    let index = 0;
+    for (let i = tree.length - 1; i >= 0; i--) {
+      const rootComment = tree[i];
+      flattenComment(rootComment, index, result);
+      index++;
+    }
+  }
+}
+
+export async function findNextUnseenVisibleCommentID(
+  mongo: MongoContext,
+  tenantID: string,
+  storyID: string,
+  userID: string,
+  orderBy: GQLCOMMENT_SORT,
+  currentCommentID?: string,
+  viewNewCount?: number
+) {
+  if (
+    ![GQLCOMMENT_SORT.CREATED_AT_ASC, GQLCOMMENT_SORT.CREATED_AT_DESC].includes(
+      orderBy
+    )
+  ) {
+    throw new Error(
+      "invalid orderBy detected: only ascending and descending order is supported."
+    );
+  }
+
+  const story = await mongo.stories().findOne({ tenantID, id: storyID });
+  if (!story) {
+    throw new StoryNotFoundError(storyID);
+  }
+
+  const user = await mongo.users().findOne({ tenantID, id: userID });
+  if (!user) {
+    throw new UserNotFoundError(userID);
+  }
+  const ignoredUserIDs = user.ignoredUsers.map((u) => u.id);
+
+  // Grab the user's seen comments collection and handle if it is null.
+  // Our dictionary/map of id -> Date will be the `seen` variable.
+  const seenComments = await findSeenComments(mongo, tenantID, {
+    storyID,
+    userID,
+  });
+  const seen = seenComments
+    ? seenComments.comments
+    : reduceCommentIDs([], new Date());
+
+  // Find a tree of only the visible comments. A comment is deemed
+  // visible if its status is in the VISIBLE_STATUSES or it is not
+  // authored by one of the user's ignored users.
+  const prunedTree = pruneTreeForVisibleComments(
+    story.tree,
+    VISIBLE_STATUSES,
+    ignoredUserIDs
+  );
+
+  // Flatten our pruned tree with only visible comments
+  const stack: FlattenedTreeComment[] = [];
+  flattenTree(prunedTree, orderBy, stack);
+
+  // Find our current position in the stack by the passed in
+  // commentID that our commenter is currently focused on
+  // with Z_KEY traversal.
+  // If currentCommentID is null, set cursor so we start at
+  // beginning of stack if order is ascending and end of stack if
+  // order is descending.
+  let cursor = 0;
+  if (!currentCommentID) {
+    if (orderBy === GQLCOMMENT_SORT.CREATED_AT_ASC) {
```

Comment:
I think the cursor should always just be set to `-1` in this case now so that we always start looking at 0 if there is no `currentCommentID`